### PR TITLE
WIP: Add in ability to parse vulns into a BOM as well as a BOM into XML with vulns

### PR DESCRIPTION
**Pull request by DarthHater**
So far so good:

- Can create a `Bom` with either vulnerabilities at the root, or attached to components (per schema, from what I understand)
- Creates XML with vulns, namespaced with `v`, and attaches NS to bom if it needs to
- Adds in a test that uses the vuln bom, parses it and checks it for valid after doing all it's work (and passes!)

### CODE REVIEW

File: src/main/java/org/cyclonedx/generators/xml/AbstractBomXmlGenerator.java [hunk 1]
```diff
@@ -184,6 +190,75 @@ private void createComponentNode(Node parent, Component component) {
...
+        }
+
+        if (vulnerability.getAdvisories() != null && !vulnerability.getAdvisories().isEmpty()) {
+            final Element advisories = createElement(vulnNode, "v:advisories");
+            for (Advisory advisory : vulnerability.getAdvisories()) {
+                final Element advisoryNode = createElement(advisories, "v:advisory");
+                advisoryNode.setTextContent(advisory.getText());
+            }
+        }
+    }
```
Comment (mr-zepol) [inline src/main/java/org/cyclonedx/generators/xml/AbstractBomXmlGenerator.java:260]:
```suggestion
               final Element ratingNode = createElement(ratings, "v:rating");
                final Element score = createElement(ratingNode, "v:score");

                createNodeElement(score, "v:base", rating.getScore().getBase());
                createNodeElement(score, "v:impact", rating.getScore().getImpact());
                createNodeElement(score, "v:exploitability", rating.getScore().getExploitability());
                createNodeElement(ratingNode, "v:severity", rating.getSeverity());
                createNodeElement(ratingNode, "v:method", rating.getMethod());
                createNodeElement(ratingNode, "v:vector", rating.getVector());
            }
        }

        if (vulnerability.getCwes() != null && !vulnerability.getCwes().isEmpty()) {
            final Element cwes = createElement(vulnNode, "v:cwes");
            for (Cwe cwe : vulnerability.getCwes()) {
                createNodeElement(cwes, "v:cwe", cwe.getText());
            }
        }

        if (vulnerability.getRecommendations() != null && !vulnerability.getRecommendations().isEmpty()) {
            final Element recommendations = createElement(vulnNode, "v:recommendations");
            for (Recommendation recommendation : vulnerability.getRecommendations()) {
                createNodeElement(recommendations, "v:recommendation", recommendation.getText());
            }
        }

        if (vulnerability.getAdvisories() != null && !vulnerability.getAdvisories().isEmpty()) {
            final Element advisories = createElement(vulnNode, "v:advisories");
            for (Advisory advisory : vulnerability.getAdvisories()) {
                createNodeElement(advisories, "v:advisory", advisory.getText());
            }
        }
    }

    private void createNodeElement(final Element mainElement, String nameElement, String elementText) {
        final Element elementNode = createElement(mainElement, nameElement);
        elementNode.setTextContent(elementText);
    }
```

Comment (mr-zepol) [reply src/main/java/org/cyclonedx/generators/xml/AbstractBomXmlGenerator.java:260]: It creates a method to remove duplicated code

File: src/main/java/org/cyclonedx/generators/json/AbstractBomJsonGenerator.java [hunk 2]
```diff
@@ -125,9 +133,77 @@ JSONObject createComponentNode(final Component component) {
...
+            }
+            parent.put("vulnerabilities", jsonVulns);
+        }
+    }
+
+    private void createVulnerabilityNode(final JSONArray parent, Vulnerability1_0 vulnerability) {
+        final JSONObject jsonVuln = OrderedJSONObjectFactory.create();
+        jsonVuln.put("ref", vulnerability.getRef());
+        jsonVuln.put("id", vulnerability.getId());
+        jsonVuln.put("description", vulnerability.getDescription());
```
Comment (kakumara) [inline src/main/java/org/cyclonedx/generators/json/AbstractBomJsonGenerator.java:156]: What about `source` ? https://github.com/CycloneDX/specification/blob/master/schema/ext/vulnerability-1.0.xsd#L179

File: src/main/java/org/cyclonedx/generators/xml/AbstractBomXmlGenerator.java [hunk 3]
```diff
@@ -184,6 +190,72 @@ private void createComponentNode(Node parent, Component component) {
...
+        if (vulnerability.getId() != "") {
+            final Element id = createElement(vulnNode, "v:id");
+            id.setTextContent(vulnerability.getId());
+        }
+
+        if (vulnerability.getDescription() != "") {
+            final Element description = createElement(vulnNode, "v:description");
+            description.setTextContent(vulnerability.getDescription());
+        }
+
```
Comment (kakumara) [inline src/main/java/org/cyclonedx/generators/xml/AbstractBomXmlGenerator.java:216]: similar to above, what about `source` ? https://github.com/CycloneDX/specification/blob/master/schema/ext/vulnerability-1.0.xsd#L179

File: src/test/java/org/cyclonedx/BomXmlGeneratorTest.java [hunk 4]
```diff
@@ -89,6 +89,17 @@ public void schema11WithDependencyGraphGenerationTest() throws Exception {
         Assert.assertTrue(parser.isValid(file, CycloneDxSchema.Version.VERSION_11));
     }
 
+    @Test
+    public void schema11WithVulnerabilityTest() throws Exception {
```
Comment (kakumara) [inline src/test/java/org/cyclonedx/BomXmlGeneratorTest.java:93]: Q: should we have a test for schema12 here as well? (with the same vuln xml)

Comment (bhamail) [reply src/test/java/org/cyclonedx/BomXmlGeneratorTest.java:94]: > Q: should we have a test for schema12 here as well? (with the same vuln xml)

Makes sense to me. Added: b020ca1

File: .github/workflows/maven.yml [hunk 5]
```diff
@@ -9,6 +9,8 @@ jobs:
     strategy:
       matrix:
         java: [8, 11]
+      # Run all tests even if one fails
+      fail-fast: false
```
Comment (DarthHater) [inline .github/workflows/maven.yml:13]: NICE

File: src/main/java/org/cyclonedx/generators/json/AbstractBomJsonGenerator.java (hunk 2 above)
Comment (bhamail) [reply src/main/java/org/cyclonedx/generators/json/AbstractBomJsonGenerator.java]: outdated now

File: src/main/java/org/cyclonedx/generators/xml/AbstractBomXmlGenerator.java (hunk 3 above)
Comment (bhamail) [reply src/main/java/org/cyclonedx/generators/xml/AbstractBomXmlGenerator.java]: outdated now

File: src/main/java/org/cyclonedx/CycloneDxSchema.java [hunk 6]
```diff
@@ -153,22 +152,34 @@ public Schema getXmlSchema(InputStream... inputStreams) throws SAXException {
...
-                .schemaJson(rawSchema)
-                .build();
-        return schemaLoader.load().build();
+    private org.everit.json.schema.Schema getJsonSchema(InputStream bomSchemaStream, InputStream... extSchemaStreams) throws URISyntaxException {
+        final JSONObject spdxSchema = new JSONObject(new JSONTokener(
+            this.getClass().getClassLoader().getResourceAsStream("spdx.schema.json")));
+        final SchemaLoaderBuilder schemaLoaderBuilder = SchemaLoader.builder()
+            .registerSchemaByURI(new URI(spdxSchema.get("$id").toString()), spdxSchema);
+
+        for (int i=0; i<extSchemaStreams.length; i++) {
```
Comment (kakumara) [inline src/main/java/org/cyclonedx/CycloneDxSchema.java:177]: minor: not sure how strict this repo adhere to formatting, but the formatting (spacing) here  doesn't look correct to me

File: src/main/java/org/cyclonedx/converters/ExtensionAdapter.java [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+import org.cyclonedx.model.vulnerability.Vulnerability1_0;
+
+public class ExtensionAdapter
+    extends TypeAdapter<Map<String, Extension>>
+{
+  @Override
+  public void write(
+      final JsonWriter jsonWriter, final Map<String, Extension> stringExtensionMap) throws IOException
+  {
+
```
Comment (kakumara) [inline src/main/java/org/cyclonedx/converters/ExtensionAdapter.java:27]: minor: I would put a comment here (usually something like `// noop`)

File: src/main/java/org/cyclonedx/converters/ExtensionConverter.java [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+    public ExtensionConverter(final Class type,
+                              final ExtensionType extensionType)
+    {
+        this.type = type;
+        this.extensionType = extensionType;
+    }
+
+    @Override
+    public void marshal(Object object, HierarchicalStreamWriter writer, MarshallingContext context) {
+
```
Comment (kakumara) [inline src/main/java/org/cyclonedx/converters/ExtensionConverter.java:50]: ditto above https://github.com/CycloneDX/cyclonedx-core-java/pull/51/files#r551997584

File: src/main/java/org/cyclonedx/generators/json/AbstractBomJsonGenerator.java [hunk 9]
```diff
@@ -94,6 +108,109 @@ void createComponentsNode(final JSONObject json, final List<Component> component
...
+                }
+            }
+        }
+    }
+
+    private void processVulnerabilities(final JSONObject json, final Extension extension, final String nodeName){
+        if (extension != null && extension.getExtensions() != null && !extension.getExtensions().isEmpty()) {
+            final JSONArray vulnArray = new JSONArray();
+            for (ExtensibleType ext : extension.getExtensions()) {
+                final Vulnerability1_0 vuln = (Vulnerability1_0) ext;
```
Comment (kakumara) [inline src/main/java/org/cyclonedx/generators/json/AbstractBomJsonGenerator.java:125]: Q: should we check if the `ext` is an `instanceof Vulnerability1_0`  before casting here?

File: src/main/java/org/cyclonedx/parsers/JsonParser.java [hunk 10]
```diff
@@ -62,7 +63,9 @@ public Bom parse(final File file) throws ParseException {
      */
     public Bom parse(final byte[] bomBytes) throws ParseException {
         try {
-            return JSON.parseObject(bomBytes, Bom.class, Feature.SupportNonPublicField);
+            GsonBuilder builder = new GsonBuilder();
+            Gson gson = builder.create();
```
Comment (kakumara) [inline src/main/java/org/cyclonedx/parsers/JsonParser.java:67]: Q: why is the use of builder here? (while in the other occurrences  just using `new Gson()`)

**Comment by DarthHater** [comment]
I'm closing this as we've largely moved to #61 , and the following three PRs on my fork: https://github.com/DarthHater/cyclonedx-core-java/pull/2
https://github.com/DarthHater/cyclonedx-core-java/pull/3
https://github.com/DarthHater/cyclonedx-core-java/pull/4

Those tell the entire Jackson story, more or less!

**Pull request closed without merging**